ibis/pandas/udf.py

```diff
@@ -0,0 +1,192 @@
+import ibis.expr.operations as ops
+
+from ibis.pandas.core import scalar_types
+from ibis.pandas.dispatch import execute_node, Dispatcher, pause_ordering
+
+
+rule_to_python_type = Dispatcher('rule_to_python_type')
+
+
+def nullable(datatype):
+    return (type(None),) if datatype.nullable is True else ()
+
+
+with pause_ordering():
+
+    @rule_to_python_type.register(dt.DataType)
+    def datatype_rule(rule, groupby=False):
+        return scalar_types
+
+    @rule_to_python_type.register(dt.Array)
+    def array_rule(rule, groupby=False):
+        return (list,)
+
+    @rule_to_python_type.register(dt.Map)
+    def map_rule(rule, groupby=False):
+        return (dict,)
+
+    @rule_to_python_type.register(dt.Struct)
+    def struct_rule(rule, groupby=False):
+        return (collections.OrderedDict,)
+
+    @rule_to_python_type.register(dt.String)
+    def string_rule(rule, groupby=False):
+        return six.string_types
+
+    @rule_to_python_type.register(dt.Integer)
+    def int_rule(rule, groupby=False):
+        return six.integer_types + (np.integer,)
+
+    @rule_to_python_type.register(dt.Floating)
+    def float_rule(rule, groupby=False):
+        return (float, np.floating)
+
+
+def signature(input_type, klass):
+    return (
+        (klass,) + rule_to_python_type(r) + nullable(r) for r in input_type
+    )
+
+
+def udf(input_type, output_type):
+    """Define a UDF (user-defined function) that operates element wise on a
+    Pandas Series.
+
+    Parameters
+    ----------
+    input_type : List[ibis.rules]
+        A rules encoding the abstract type of each argument. These are found in
+        :mod:`~ibis.expr.rules`
+    output_type : ibis.rules or ibis.expr.datatypes.DataType
+        The abstract return type of the function
+
+    Examples
+    --------
+    >>> import ibis
+    >>> @udf(input_type=[dt.string], output_type=dt.int64)
+    ... def my_string_length(series, **kwargs):
+    ...     return series.str.len() * 2
+    """
+    def wrapper(func):
+        # generate a new custom node
+
+        UDFNode = type(
+            func.__name__ + '_udf_node',
+            (ops.ValueOp,),
+            dict(input_type=input_type, output_type=output_type.array_type)
+        )
+
+        with pause_ordering():
+            @execute_node.register(
+                UDFNode, *signature(input_type, klass=pd.Series)
+            )
+            def execute_udf_node(op, *args, **kwargs):
+                return func(*args, **kwargs)
+
+            @execute_node.register(
+                UDFNode, *signature(input_type, klass=SeriesGroupBy)
+            )
+            def execute_udf_node_groupby(op, *args, **kwargs):
+                groupers = [
+                    grouper for grouper in (
+                        getattr(arg, 'grouper', None) for arg in args
+                    ) if grouper is not None
+                ]
+
+                assert all(groupers[0] == grouper for grouper in groupers[1:])
```

Comment:
when would this fail?

Reply:
It should never fail, that's why it's an assertion and not a user facing error message.